std/assembly/math.ts

```diff
@@ -2287,6 +2287,114 @@ export namespace NativeMathf {
   }
 }
 
+export namespace SafeMath {
+
+  @inline
+  export function clz<T>(x: T): T {
+    if (isInteger<T>()) {
+      return builtin_clz<T>(x);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function clz32<T>(x: T): T {
+    if (isInteger<T>() && sizeof<T>() <= 4) {
+      return builtin_clz<T>(x);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function min<T>(x: T, y: T): T {
+    if (isInteger<T>()) {
+      return builtin_min<T>(x, y);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function max<T>(x: T, y: T): T {
+    if (isInteger<T>()) {
+      return builtin_max<T>(x, y);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function abs<T>(x: T): T {
+    if (isInteger<T>()) {
+      return builtin_abs<T>(x);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function imul<T>(x: T, y: T): T {
+    if (isInteger<T>()) {
+      return x * y;
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function pow<T>(x: T, y: i32): T {
+    if (isInteger<T>()) {
+      if (sizeof<T>() == 8) return <T>ipow64(<i64>x, y);
+      if (sizeof<T>() <= 4) return <T>ipow32(<i32>x, y);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function sign<T>(x: T): T {
+    if (isInteger<T>()) {
+      if (!x) return <T>0;
```

Comment:
isn't this short-circuit only needed for unsigned case?

Reply:
Current logic follow this scenario:
1) if `T` is **unsigned** that result will be `1` for all values except `0` which in this case also would `0`.
2) if `T` is **signed** that result will be `-1` for `x < 0`, `1` for `x > 0` and `0` for `0`.

Reply:
Do you mean that `1 - (x >> (sizeof<T>() * 4 - 1) << 1)` doesn't work for 0?

Reply:
This line of code defines the highest bit and leads to the range [-1, 1] but it can't determine zero from positive value

Reply:
nvm then